lib/iris/experimental/stratify.py

```diff
@@ -0,0 +1,179 @@
+
+    This routine is especially useful for computing isosurfaces of
+    phenomenon that are generally monotonic in the direction of interpolation
+    (such as height/pressure and salinity/depth).
+
+    Parameters
+    ----------
+    phenom : :class:`iris.cube.Cube`
+        The data to be re-levelled.
+
+    level_data : :class:`~iris.cube.Cube` or :class:`~iris.coord.Coord`
+        The coordinate values (cube/coord) in the same system as
+        ``target levels``.
+        All dimensions of ``level_data`` must be broadcastable to ``phenom``
+
+    target_levels : array-like
+        The levels of ``level_data`` to pick from ``phenom``.
+        For the default interpolator, this must be a 1d array.
+        For a custom interpolator this may be multi-dimensional and must
+        broadcast to phenom and level_data except in the phenomenon's
+        ``interp_direction`` dimension.
+
+    interp_direction : int or :class:`~iris.coords.Coord` instance
+        The direction of interpolation. This is necessary because
+        level_data is rarely 1D. This is often the z dimension coordinate
+        of ``phenom``.
+
+    interpolator : callable or None
+        The interpolator to use when computing the interpolation. The function
+        will be passed the following positional arguments::
+
+            (target levels, level data, phenom data, interp axis)
+
+        If the interpolator is None, :func:`stratify.interpolate` will be used
+        with linear interpolation and NaN extrapolation.
+
+        An example of constructing an alternative interpolation scheme:
+
+            from functools import partial
+            interpolator = partial(stratify.interpolate,
+                                   interpolation=stratify.INTERPOLATE_NEAREST,
+                                   extrapolation=stratify.EXTRAPOLATE_LINEAR)
+
+    """
+    # Identify the z-coordinate within the phenom cube
+    if isinstance(interp_direction, Coord):
+        [interp_dim] = phenom.coord_dims(interp_direction)
+    else:
+        interp_dim = interp_direction
+
+    if isinstance(level_data, Coord):
+        source_data = level_data.points
+    else:
+        source_data = level_data.data
+
+    # phenom and level_data must be broadcastable.
+    phenom_data, source_data = np.broadcast_arrays(phenom.data,
+                                                   source_data)
+
+    target_levels = np.array(target_levels)
+    target_phenom_dim = interp_dim
+    if target_levels.ndim != 1:
+        # target levels must be broadcastable to phenom in everything
+        # but the interpolation axis (if not, raise)
+        t_dim_delta = phenom_data.ndim - target_levels.ndim
+        tl_axis = interp_dim - t_dim_delta
+        data_shape = list(phenom_data.shape)
+        data_shape.pop(interp_dim)
+        target_shape = list(target_levels.shape)
+        target_shape.pop(tl_axis)
+        np.broadcast_arrays(np.empty(data_shape), np.empty(target_shape))
```

Comment:
@pelson A `try ... except` around the `broadcast_arrays` would be helpful in user-land when it barfs.